Reminder-App/Reminder-App/ReminderListViewController.swift

```diff
@@ -9,32 +9,217 @@ import UIKit
+                            
+                            let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
+                            let actionTitle = NSLocalizedString(ErrorMessages.okMessage, comment: ErrorMessages.okActionTitle)
+                            
+                            alert.addAction(UIAlertAction(title: actionTitle,
+                                                          style: .default,
+                                                          handler: {
+                                                            _ in
+                                                            self.dismiss(animated: true, completion: nil)
+                                                          }))
+                            
+                            self.present(alert, animated: true, completion: nil)
+                        }
+                    }
+                }
+            })
         }
     }
     
     // MARK: Lifecycle Methods
     override func viewDidLoad() {
-        
         super.viewDidLoad()
         
-        reminderListDataSource = ReminderListDataSource()
+        reminderListDataSource = ReminderListDataSource(
+            reminderCompletedAction: {
+                reminderIndex in
+                DispatchQueue.main.async {
+                    self.tableView.reloadRows(at: [IndexPath(row: reminderIndex, section: 0)], with: .automatic)
+                    self.refreshProgressView()
+                }
+            }, reminderDeletedAction: {
+                DispatchQueue.main.async {
+                    self.refreshProgressView()
+                }
+            }, remindersChangedAction: {
+                DispatchQueue.main.async {
+                    self.tableView.reloadData()
+                    self.refreshProgressView()
+                }
+            })
+        
         tableView.dataSource = reminderListDataSource
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        let radius = view.bounds.size.width * 0.5 * 0.7
+        progressContainerView.layer.cornerRadius = radius
+        progressContainerView.layer.masksToBounds = true
+        self.refreshProgressView()
+        refreshBackground()
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        if let navigationController = navigationController,
+           navigationController.isToolbarHidden {
+            navigationController.setToolbarHidden(false, animated: animated)
+        }
+    }
+    
+    // MARK: Private Methods
+    private func addReminder() {
+        let storyboard = UIStoryboard(name: Self.mainStoryboardName, bundle: nil)
+        let detailViewController: ReminderDetailViewController = storyboard.instantiateViewController(identifier: Self.detailViewControllerIdentifier)
```

Comment:
use self only where its required
like in in initializer where parameter have same name.
or in Closures

Reply:
It is a static property so it can be accessed by either Struct's name or by Self. The struct's name is lengthy so thats why using Self.